### ISSUE TITLE
Previously presisted `isDeleted` child records cause parent to be invalid

### ISSUE THREAD

**Issue by Rio517** [offirgolan/ember-cp-validations#205]
### tldr; 

In the case where a parent `has-many` children, if deleting a previously persisted, invalid child record, the parent never seems to return to a valid state.  

### Background
For example:

1. A Parent and two children (let's call them C1 and C2) are loaded from the data store
2. A field on C1 is made invalid
3. C1 is deleted in the client
4. Parent is now invalid and validations prevent it from being persisted again. I find my parentModel is in the following state:

```js
parentModel.get('validations.attrs.childModels.messages')
["This field can't be blank"]
parentModel.get('validations.messages')
["This field can't be blank"]
```

`parentModel.get('childModels.length')` is still 2, but `parentModel.get('firstObject.isDeleted')` is also true.

### Basic Model Setup

```js
// parentModel.js 
const Validations = buildValidations({
  childModels: validator('has-many')
});

//  child-model.js
const Validations = buildValidations({
  name: validator('presence', true),
  phone: validator('presence', true),
  role: validator('presence', true),
});
```

### What I tried:

I thought I might work around this with a disabled function, but `ember-cp-validations` seems smart enough to disable the validation already and `disabled()` never actually gets called.

```js
const Validations = buildValidations(
  {...},{
    disabled() {
      console.log(this.get('model.isDeleted'));
      return this.get('model.isDeleted');
    }
  });
```

I was also trying to traverse down the validations.attrs.childModels tree to see if a copy of the isDeleted record was somehow in there, but I couldn't quite find it.  I'm still fairly new to Ember.

Any help appreciated!

P.S. This library is great, the documentation is also very good, which is really appreciated.

**Comment by offirgolan** [comment]
Have you tried unloading the record from the store after delete or just removing it from the hasMany array?

```js
child.destroy().then(() => {
 this.get('store').unloadRecord(child);
 parent.get('childModels').removeObject(child); // might not be needed 
});
```

**Comment by Rio517** [comment]
While this workaround works for the validation state, it prevents persisting the delete record when saving the parent.  

For example, in my case, I'm calling `deleteRecord` and only later calling `save` on the parent.  In that case, the chilren never get deleted on the server.

**Comment by offirgolan** [comment]
When you added the `disabled` option, did you also add `dependentKeys: ['isDeleted']`? Im assuming thats why disabled didnt get called

**Comment by offirgolan** [comment]
@Rio517 did my pervious solution solve your issue?

**Comment by Rio517** [comment]
Hi @offirgolan !  Sorry, I missed your other comment. Github didn't notify me.  Anyway... 

Just to be clear: the `disabled()` did get called prior to deleting the child record. I would see "false" being logged to console. After the child record was deleted, the `disabled()` was no longer called.

I'm not clear on how the `dependentKeys` should be passed in. I would have guessed I'd need to add this to every field on the childModel, but that hasn't seemed to help.  So now my child model validation looks like so:

```js
const Validations = buildValidations({
  name: validator('presence', true, {dependentKeys: ['isDeleted'] }),
  phone: validator('presence', true, {dependentKeys: ['isDeleted'] }),
  role: validator('presence', true, {dependentKeys: ['isDeleted'] }),
},{
  disabled() {
    console.log(this.get('model.isDeleted'));
    return this.get('model.isDeleted');
  }
});
```

Am I understanding this correctly?  The only reference I could find to how the library was handling `dependentKeys` was [here](https://github.com/offirgolan/ember-cp-validations/blob/e07783823fbd74f12c253411e74b6cbabeda4e63/addon/validations/validator.js#L33).

**Comment by offirgolan** [comment]
@Rio517 A validator excepts only 2 arguments, the type, and the options. To do what you are looking for, the validator will look like this

`validator('presence', { presence: true, dependentKeys: ['isDeleted'] })`

Or... you can place dependentKeys in the global options space as well.

```js
const Validations = buildValidations({
  name: validator('presence', true),
  phone: validator('presence', true),
  role: validator('presence', true)
},{
  dependentKeys: ['isDeleted'],
  disabled() {
    console.log(this.get('model.isDeleted'));
    return this.get('model.isDeleted');
  }
});
```

**Comment by Rio517** [comment]
Prefect.  I didn't realize I could set global options like that.   All working now.  Thanks for all the help!

**Comment by offirgolan** [comment]
Fantastic! 😺
